### PR TITLE
Add new workflow enabling DPL level record file creation for muon alignment

### PR DESCRIPTION
**Pull request by Elros60**
This is a proposal of having a DPL-level task for creating `MillePede2` record file to be used by muon alignment. Some adaptations have been made as well in `Detectors/ForwardAlign/` especially to disable the internal record writer of aligner class. Ideally this workflow should be placed right after matching of `MCH-MID`, tests have been done locally with minimum modifications in `async_pass.sh` fetched from O2DPG.

### CODE REVIEW

File: Detectors/ForwardAlign/src/MillePede2.cxx
Comment (pillot) [inline Detectors/ForwardAlign/src/MillePede2.cxx]: There is a memory leak here. If the recording is disabled, shouldn't fRecord remains a nullptr and simply be tested not to use it in the code ?

File: Detectors/MUON/MCH/Align/CMakeLists.txt
Comment (pillot) [inline Detectors/MUON/MCH/Align/CMakeLists.txt]: most of the libraries are linked to MCHAlign, so you probably need only this one here ?

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: There is a memory leak. But you don't need to copy the clusters, you could simply build the track from the input clusters, which remain in memory during the entire processing ?

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: max chi2 ~ infinity. If you do not need to improve the track (i.e. remove some clusters during the refitting), it is enough to refit the track once with the smoother enabled to recompute the parameters at each cluster: `trackFitter.fit(track);`

File: Detectors/MUON/MCH/Align/include/MCHAlign/Aligner.h
Comment (pillot) [inline Detectors/MUON/MCH/Align/include/MCHAlign/Aligner.h]: not sure how big is this object, but is it really necessary to return it by value? (By the way returning a (const) reference is usually better than a pointer.)

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx:119]: This will use the default chamber resolution. Is it what you want?

File: Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx:234]: The cluster resolution will be used (while the default chamber resolution is used in AlignRecordSpec). It it what you want?

File: Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx]: same comment as above. A simple fit will do the same in this case.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: This test can be performed using the `mchTrack`, before producing the `convertedTrack`.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: not sure if/how this work, but in principle this should be done once per TF (i.e. per `run(...)`), probably sending one vector of `MillePedeRecord` in this case.

**Review by pillot — COMMENTED** [review]
Hi Chi, thanks for the PR. Is the manual reading of the track files in AlignmentSpec also expected to be replaced with this new workflow ?

**Comment by Elros60** [comment]
> Hi Chi, thanks for the PR. Is the manual reading of the track files in AlignmentSpec also expected to be replaced with this new workflow ?

Hi Philippe,
Thanks a lot for your comment. So let me first explain the current situation related to my PR.
Actually, there were other in-MillePede function for record reading/writing made by Andry(`MilleRecordWriter` and `MilleRecordReader`), but these are not quite adequate for DPL level usage. So I just made another optional task for record writing adapted with DPL requirements(there should be a similar task for MFT as well); to use this one I need to disable the internal I/O of records that are currently coupled inside `MillePede2` class. The DPL level record I/O implementation is not yet ready for MFT at this moment, Andy is working on it actually and meanwhile she will still need the internal I/O to test codes. So I think for now, it's fine to keep both: my DPL task for MCH record and internal I/O for MFT testing.

Concerning the `AlignmentSpec`, it was made to be able to handle all the cases we've met before including the option using `mchtracks.root`, and at this moment it remains an executable not a real "workflow". So for muon alignment usage, I would suggest to keep it local and not bring it into DPL; i.e. we will only make `AlignRecordSpec` available for DPL and once we obtain records file from reconstruction, one then only needs to run locally the alignment workflow to get alignment parameters.

File: Detectors/ForwardAlign/src/MillePede2.cxx
Comment (Elros60) [reply Detectors/ForwardAlign/src/MillePede2.cxx]: This was actually a difficulty that I found when I try to decouple RecordWriter from MillePede2. I managed to find an easy way to get rid of this. I will push my modifications for this.

File: Detectors/MUON/MCH/Align/CMakeLists.txt
Comment (Elros60) [reply Detectors/MUON/MCH/Align/CMakeLists.txt]: This is fixed. Only `O2::MCHAlign` will be kept.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: Corrected according to your suggestion

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: I actually need this improvement for tracks, and what I made as change is to remove the manual setting for improve cut value so that it will be configured with `dpl-workflow.sh` with configurations key values like `MCHTracking.chamberResolutionX=0.4;MCHTracking.chamberResolutionY=0.4;MCHTracking.sigmaCutForTracking=7.;MCHDigitFilter.timeOffset=126;MCHTracking.sigmaCutForImprovement=6.;MCHTracking.maxCandidates=20000;`. With this, I think, the workflow will be able to get informed with those input parameters, am I right?

File: Detectors/MUON/MCH/Align/include/MCHAlign/Aligner.h
Comment (Elros60) [reply Detectors/MUON/MCH/Align/include/MCHAlign/Aligner.h]: This is modified.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx:119]: Same for the previous comment, the workflow will be configured with key values containing settings for trackfitter `MCHTracking.chamberResolutionX=0.4;MCHTracking.chamberResolutionY=0.4;MCHTracking.sigmaCutForTracking=7.;MCHDigitFilter.timeOffset=126;MCHTracking.sigmaCutForImprovement=6.;MCHTracking.maxCandidates=20000;`

File: Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx:234]: This is a mistake I made when I was doing formatting for PR preparation. This is fixed

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: This actually works for my case; before I arrived at this solution I also tested with a vector valued branch like what you have done for MCH tracks and clusters. But for the case of alignment, it seems to me more convenient to save directly a TTree of records so MillePede2 can directly read and use it. For the above one, from the test I've done with 3 CTFs, I actually didn't capture any trouble.
I'm not sure if I understood correctly: by one vector you mean I keep the current definition of branch but instead of sending records one by one I save them in a vector when doing loop then send the resulting vector directly to output?

Comment (shahor02) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: I am also not sure how this will work, perhaps @ktf can comment if creating multiple snapshots of the same output in a single TF will not lead to only last one or 1st one actually received by the consumer. But for sure, if in the TF you don't produce an output the DPL will warn you about the lost TF, since you define the `{"MUON", "RECORD_MCHMID", 0}` with `Lifetime::Timeframe`. If the output is sporadic, you better define https://github.com/AliceO2Group/AliceO2/blob/5d3fdf914b7208018a78f6500d9325ee21c06caf/Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx#L346 with `Lifetime::Sporadic`.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: this is not used anymore

File: Detectors/MUON/MCH/Align/src/Aligner.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/Aligner.cxx]: These deletes should go in the destructor.
Actually, since these pointers point to objects produced only once in the constructor, you could use data member objects instead of pointers. If you don't want to change the others, at least `fTrackRecord` can remain a data member object as before.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: Ok, I see, but `trackFitter.getMaxChi2()` is not what you want then. The parameter you need to set the chi2 cut is `sigmaCutForImprovement` and you can access it via `TrackerParam::Instance()`. You can have a look at the `TrackFinder` class, and in particular at the functions `init()`and `improveTracks()` for a concrete example.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx:119]: Also here you need to setup the chamber resolution in the `TrackFitter` taking the parameters from `TrackerParam::Instance()` as it is done in `TrackFinder::init()`:
`mTrackFitter.setChamberResolution(trackerParam.chamberResolutionX, trackerParam.chamberResolutionY);`

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: This is fixed. Thank you!

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (shahor02) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: I asked @ktf , the DPL in this case will create a multip-part message with each snapshot being a single part. On the receiving side you would need to explicitly do something like
```
auto pos = inputs.getPos(binding);
size_t nparts = inputs.getNofParts(pos);
for (size_t i=0;I<nparts;i++) {
   auto mpRecord = inputs.get<fwdalign::MillePedeRecord>(binding, I);
   // ... use the record
}
```
With the default `inputs.get<>(binding)` you will get only 1st out of potentially many tracks. I think this is what RootTreeWriter is using. You should probably check if the number of tracks you write in this device is equal to those in the written file.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: Ok, thanks a lot! I'm making modifications!

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx:119]: Thanks a lot!

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: By looking at your example of `TrackFinder`, I'm trying to do the same thing as you've done in `TrackFinder`; but there is one thing I'm not quite sure: if `trackFitter.setBendingVertexDispersion(trackerParam.bendingVertexDispersion);` is necessary for our case?

Comment (pillot) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: This parameter is indeed used to initiate the refit. To be consistent it might be better to set it as well (although the default value in TrackFitter is equal to the default value in TrackerParam so it will not change the results if you keep the default).

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: Thanks a lot for info!
With my current version of sending tracks to output(using `LifeTime::TimeFrame`), I've got actually a total number of entries for my output record file, that is compatible with what I've got using AO2D. And the content inside this root file seems fine to me as well cuz I can actually run my alignment workflow with this record file as input, and I didn't get any trouble...

I will check if the number of tracks that are being processed corresponds to the number of entries of my output record file. I will let you know very shortly.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: I see. Thanks a lot for explanation!

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: HI,
I tested already with `Lifetime::Sporadic`: I actually have ~40 tracks candidates per `run`, and the number of total tracks candidates that were being converted into records corresponds to the number of total entries I have in my output ROOT file of `MillePede` records. I may stay with this solution which leads me to the smallest modifications compared to the solution proposed by Giulio.

**Comment by shahor02** [comment]
Hi,
The Lifetime::Sporadic should be used anyway since the output in every TF is not guaranteed. Concerning the possible data loss due to the multiple snapshots per TF, this could happen only in TFs where there are more than 1 tracks converted.

**Comment by Elros60** [comment]
> Hi, The Lifetime::Sporadic should be used anyway since the output in every TF is not guaranteed. Concerning the possible data loss due to the multiple snapshots per TF, this could happen only in TFs where there are more than 1 tracks converted.


Thanks a lot for info!
Exactly, and in my case, I have ~40 tracks need to be converted per TF and they are all saved correctly into ROOT file. I tested with one CTF, I had a total number of **2233** MCH-MID tracks to be converted and I received **2233** entries in my record ROOT file. So at this stage, I don't see any data loss using only `Sporadic`.
Then to validate these records at alignment level with more data, I still need one more test with 3 CTFs to do a one-to-one check with one AO2D from official reconstruction.

**Comment by shahor02** [comment]
OK, good, thanks for checks. It might be that the RootTreeWriter does actually loop over the sub-parts of the multipart message.

**Comment by Elros60** [comment]
> OK, good, thanks for checks. It might be that the RootTreeWriter does actually loop over the sub-parts of the multipart message.

Thank you so much!
Final test will take 10 hours on my pc(I need to include TPC as well for my full local reco), I will finalise this PR by tomorrow morning. Thanks again for all information related to DPL and all your comments!

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: the chi2 cut is defined as:
`mImproveCutChi2 = 2. * trackerParam.sigmaCutForImprovement * trackerParam.sigmaCutForImprovement;`

File: Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx]: same here:
mImproveCutChi2 = 2. * trackerParam.sigmaCutForImprovement * trackerParam.sigmaCutForImprovement;

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (pillot) [inline Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: this is a detail, but currently `mTrackCount` and `mTrackMatchedPerTF` counters are identical.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: Thanks for pointing this out.

File: Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx
Comment (Elros60) [reply Detectors/MUON/MCH/Align/src/AlignRecordSpec.cxx]: Yes, indeed, another one was meant to check per TF track candidates, then I made some changes; I should have removed it.

**Comment by Elros60** [comment]
Test with 3 CTFs comparing with official AO2D has been done this morning. Result with alignment workflow using produced records looks fine for me. A more relevant test of consistency with whole official reconstruction will be possible once this PR is merged into `dev` and this record workflow is included in reconstruction.

Thanks a lot for all comments and suggestions for fix from Ruben and Philippe!

**Review by pillot — APPROVED** [review]
Alright, then it looks ok on my side. Thank you!

**Comment by Elros60** [comment]
Dear @shahor02,
I've got the approval from Philippe.
I would just kindly ask if everything fine for your side?

**Review by shahor02 — APPROVED** [review]

**Pull request merged**
